### ISSUE TITLE
"KeyStoreException: the master key exists but is unusable" while trying to connect to the Tap to Pay on specific Samsung devices

### ISSUE THREAD

**Issue by wkol** [stripe/stripe-terminal-android#444]
## Summary
Since the release of our app that integrates the Tap to Pay feature, we have received multiple crash reports on Firebase Crashlytics that happen during the initialization of the Terminal and connection to Tap to Pay (we connect immediately after the initialization, so it's hard to tell whether it occurs during the initialization or the connection process). All of the crashes relate to the usage of the (I believe) EncryptedSharedPreferences and share a common exception:
```
Unable to bind to service com.stripe.cots.aidlservice.CotsService with Intent { act= cmp=<our-app-name>/com.stripe.cots.aidlservice.CotsService }:
java.security.KeyStoreException: the master key android-keystore://asp_key exists but is unusable
```
What caught my attention was that all those crashes so far (over 60 during the last five days) have happened on the various models of Samsung S24 (either S24, S24+, or S24 Ultra) - all running Android 14.
During the research, [I found an issue in the Tink-Crypto repository on GitHub](https://github.com/tink-crypto/tink/issues/535) that describes an identical error. In this issue, [one of the contributors states](https://github.com/tink-crypto/tink/issues/535#issuecomment-912170221):
> We've found that Android Keystore occasionally corrupts the master key on certain devices. We don't know why, we think it could be due to faulty OEM firmware/hardware. 

Based on the fact that we have over 1000 Terminal objects with `device_type="mobile_phone_reader"` and only 60 crashes (that happened solely on Samsung S24 devices), I believe the cause is the same as described in the issue. Nevertheless, a workaround is suggested for the issue, which includes destroying the key and the storage in case of such errors. 

It would be great if you could take a look at this and implement the workaround if you decide that this may be indeed the problem. I believe we will encounter more and more crashes like this in our app. If you need more details, just let me know.
## Code to reproduce
```kotlin
Terminal.initTerminal(
    // ...
)
if (Terminal.isInitalized()) {
    Terminal.getInstance().discoverReaders(
        config = DiscoveryConfiguration(
            discoveryMethod = DiscoveryMethod.LOCAL_MOBILE,
            //...
        ),
        discoveryListener = object : DiscoveryListener {
            override fun onUpdateDiscoveredReaders(readers: List<Reader>) {
                readers.firstOrNull { it.deviceType == DeviceType.COTS_DEVICE }?.let { reader ->
                    Terminal.getInstance().connectLocalMobileReader(
                        reader = reader,
                        // ...
                    )
                }
            }
        },
        // ...
    )
}    

```
Also, during the examination of this crash, I found a way to reproduce this crash. It involves manually corrupting (in this case, replacing the master key):
```kotlin
KeyStore.getInstance("AndroidKeyStore").apply {
    load(null)
    setEntry("asp_key", KeyStore.SecretKeyEntry(createKey()), null) // where createKey creates just a key with alias "asp_key"
}
```
Trying to connect to a nonsimulated LocalMobile reader after the key's corruption results in a crash in `com.stripe.cots.aidlservice` process.

## Android version
14
## Impacted devices (Android devices or readers)
Samsung S24 variants (SM-S921x, SM-S926x, SM-S928x) - Tap to Pay readers
## SDK version
2.23.3

**Comment by sidhant-stripe** [comment]
Hi @wkol! 

Thank you very much for the repro instructions, they were very helpful. I have implemented and merged a fix based on the Tink issue you linked to. The fix will be included in the `3.6.0` release, expected to come out around mid May.

Unfortunately, because 2.x SDK versions are close to EOL, we are only backporting critical fixes. As such we will not be creating a new 2.x release for this fix.

**Comment by wkol** [comment]
Well, either way, we would need to upgrade the SDK (since it's close to EOL), so that may be a reason to do it sooner than later. Anyway, thanks a lot for the time spent and the fix.
